src/config.ts

```diff
@@ -14,19 +14,23 @@ export function update<T>(key: string, value: T,
 }
 
 // Traverse a JSON value, replacing placeholders in all strings.
-function substitute<T>(val: T): T {
+async function substitute<T>(val: T): Promise<T> {
   if (typeof val === 'string') {
-    val = val.replace(/\$\{(.*?)\}/g, (match, name) => {
-      // If there's no replacement available, keep the placeholder.
-      return replacement(name) ?? match;
-    }) as unknown as T;
+    const replacementPattern = /\$\{(.*?)\}/g;
+    const replacementPromises: Promise<string | undefined>[] = [];
+    val.replace(replacementPattern, (match, name) => {
```

Comment:
This can be done a bit more efficiently, since the first time around we need only match, not replace:

```ts
const matches = val.matchAll(replacementPattern);
for (const match of matches) {
  // match[1] is the first captured group
  replacementPromises.push(replacement(match[1]));
}
```